subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolver.java

```diff
@@ -185,6 +188,13 @@ private void listVersionsForAllPatterns(List<ResourcePattern> patternList, IvyAr
         }
     }
 
+    private boolean isIncomplete(ModuleComponentIdentifier moduleComponentIdentifier) {
+        if (supportsEmptyGroupOrVersion) {
```

Comment:
❌ Do we really need a boolean here? Shouldn't we be able to tell from the Ivy pattern whether a group is necessary?

Reply:
I first thought these are independent things: What do we require? And where is it located?

As we have multiple patterns: What does it mean if one pattern requires a `organisation` and the other does not?

I guess the saves interpretation is: if something, group/organisation or version/revision, is a _required_ token in **all** of the patterns in both lists (we have two: `ivyPatterns` and `artifactPatterns`) then it is required, otherwise not.

The only problem I see is that currently we do not parse the pattern, we only substitute. We would need to add some kind of parsing for this. I'll have a look.

Reply:
There wouldn't be much parsing involved, you only need to look if it contains the string `[organisation]`. See [this](https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.repositories.IvyArtifactRepository.html#org.gradle.api.artifacts.repositories.IvyArtifactRepository) user guide page.

Regarding the different patterns: If we put the code at the point where the pattern is actually used, then there should be no question. If that specific pattern requires and organisation and we don't have one, return/fail.

Reply:
I have moved the check further down into `DefaultExternalResourceArtifactResolver` and perform it for each pattern now. The pattern "parsing" does not do a 100% accurate handling of _optional_ parts `([])` to avoid complex parsing. I think that's fine as this works for all our standard patterns.

For the maven case, an earlier check had to stay in place to avoid downloading of `maven-metadata.xml`.

The implementation can use some more cleanup. (I would like to avoid the code duplication on the three `isIncomplete()` methods. Or at least move them to one place.)

But first, I would like to have your feedback @oehme. Is this what you had in mind?

Reply:
Yes, this is what I was thinking of. We should try to do a better job of matching the optional pattern though. You can count the number of opening/closing parens before the part you are interested in. If opening>closing, you know it's optional.

Reply:
Cool. I improved the _optional_ matching and did some cleanup. I moved the _isComplete()_ check into `ResourcePattern` and call that from all the different places now.

Could you have another look at his PR?